web-server-lib/web-server/http/request.rkt

```diff
@@ -1,398 +1,719 @@
-                                            (or f00 f01)
-                                            headers
-                                            (apply bytes-append contents))])])
-                    (read-mime-multipart content-boundary in)))
-             (values
-              (delay (append (force bindings-GET) bs))
-              #f)])]
-       [else        
-        (match (headers-assq* #"Content-Length" headers)
-          [(struct header (_ value))
-           (cond
-             [(string->number (bytes->string/utf-8 value))
-              => (lambda (len)
-                   (let ([raw-bytes (read-bytes len in)])
-                     (cond
-                       [(eof-object? raw-bytes)
-                        (network-error
-                         'read-bindings
-                         "Post data ended pre-maturely")]
-                       [else
-                        (values (delay 
-                                  (append
-                                   (parse-bindings raw-bytes)
-                                   (force bindings-GET)))
-                                raw-bytes)])))]
-             [else 
-              (network-error
-               'read-bindings
-               "Post request contained a non-numeric content-length")])]
-          [#f
-           (values (delay empty) #f)])])]
-    [meth
-     (define content-type (headers-assq* #"Content-Type" headers))
-     (match (headers-assq* #"Content-Length" headers)
-       [(struct header (_ value))
-        (cond [(string->number (bytes->string/utf-8 value))
-               => (lambda (len)
-                    (let ([raw-bytes (read-bytes len in)])
-                      (cond
-                        [(eof-object? raw-bytes)
-                         (network-error
-                          'read-bindings
-                          "Post data ended pre-maturely")]
-                        [else
-                         (values (delay empty) raw-bytes)])))]
-              [else
-               (network-error 
-                'read-bindings
-                "Non-GET/POST request contained a non-numeric content-length")])]
-       [#f
-        (values (delay empty) #f)])]))
+                    [(#f #f)
+                     (network-error
+                      'reading-bindings
+                      "Couldn't extract form field name for file upload")]
+
+                    [(#f (list _ _ f0 f1))
+                     (make-binding:form (or f0 f1)
+                                        ;; FIXME
+                                        (port->bytes contents))]
+
+                    [((list _ _ f00 f01) (list _ _ f10 f11))
+                     (make-binding:file (or f10 f11)
+                                        (or f00 f01)
+                                        headers
+                                        ;; FIXME
+                                        (port->bytes contents))])])))
```

Comment:
@jeapostrophe I'm nearly done but I'm not sure how to proceed here. Uploaded files now get written to disk if they are large enough so `binding:file`'s `contents` should really just be the input port for its associated file.  Obviously, that would be a breaking change so that's problematic.  I could introduce a new binding type (say `binding:file-port`) and a flag to create `binding:file-port`s instead of `binding:file`s, but that would end up not being secure by default and it would introduce a bunch of cruft.

The best option I can think of is limiting the scope of the breaking change by renaming the `contents` field of `binding:file` to `in` and then adding a new function called `binding:file-contents` which just applies `port->bytes` to `in` and memoizes the result. I think this would be mostly backwards-compatible except in cases where someone is `match`ing on `(struct binding:file (id fname headers contents))`.

What do you think?

Reply:
> The best option I can think of is limiting the scope of the breaking change by renaming the contents field of binding:file to in and then adding a new function called binding:file-contents which just applies port->bytes to in and memoizes the result. I think this would be mostly backwards-compatible except in cases where someone is matching on (struct binding:file (id fname headers contents)).

The latest commit I pushed does this. Let me know if you want me to take a different approach.

I'm still not super happy with how the mime parser ended up (in particular, `in-http-lines` is probably not necessary and it's definitely not as safe as it could be) so I'm going to try and improve it tomorrow and add a few more test cases. I think once that's done, this'll be good to go.

Reply:
@samth thanks for the tip! I have never used impersonators before, but I was able to sketch this up pretty quickly:

```racket
#lang racket/base

(require racket/match
         racket/port)

(struct binding:file (id name headers [content #:mutable])
  #:transparent)

(struct binding:file/port binding:file (in)
  #:transparent)

(define (make-binding:file id name headers content [in #f])
  (impersonate-struct
   (binding:file/port id name headers #f (or in (open-input-string content)))
   struct:binding:file
   binding:file-content
   (let ([mem #f]
         [sem (make-semaphore 1)])
     (lambda (b c)
       (unless mem
         (call-with-semaphore sem
           (lambda _
             (unless mem
               (set! mem (port->string (binding:file/port-in b)))))))

       mem))))

(define b (make-binding:file "a" "a" null "a"))
(binding:file-content b)
(binding:file-content b)

(match b
  [(struct binding:file (id name headers content))
   content])
```

Is this along the lines of what you were thinking? It does seem like this would be more backwards-compatible than what I've done. Anything I should watch out for in particular with impersonators?

Reply:
I like this strategy for dealing with the memory limits and the impersonator trick is a nice one I would not have thought of.

Reply:
This is almost exactly what I was thinking of. One suggestion I'd make is to use the `content` field instead of `mem` -- it already has to be mutable, and that means there's less allocation. If you put the semaphore in a field as well, you can make the impersonator handler not a closure at all.